src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/CachedExpressionCompiler.cs

```diff
@@ -61,66 +69,104 @@ private static class Compiler<TModel, TResult>
                 if (_identityFunc == null)
                 {
-                    _identityFunc = expression.Compile();
+                    var identityFuncCore = expression.Compile();
+                    _identityFunc = model => identityFuncCore(model);
                 }
 
                 return _identityFunc;
             }
 
-            private static Func<TModel, TResult> CompileFromMemberAccess(
+            private static Func<TModel, object> CompileFromStaticMemberAccess(
                 Expression<Func<TModel, TResult>> expression,
                 MemberExpression memberExpression)
             {
-                // model => model.Member
+                // model => model.StaticMember
                 if (_simpleMemberAccessCache.TryGetValue(memberExpression.Member, out var result))
                 {
                     return result;
                 }
 
-                result = expression.Compile();
+                var func = expression.Compile();
+                result = model => func(model);
                 result = _simpleMemberAccessCache.GetOrAdd(memberExpression.Member, result);
+
                 return result;
             }
 
-            private static Func<TModel, TResult> CompileFromStaticMemberAccess(
+            private static Func<TModel, object> CompileFromSimpleMemberAccess(
                 Expression<Func<TModel, TResult>> expression,
                 MemberExpression memberExpression)
             {
-                // model => model.StaticMember
+                // Input: () => m.Member
+                // Output: () => (m == null) ? null : m.Member
+
                 if (_simpleMemberAccessCache.TryGetValue(memberExpression.Member, out var result))
                 {
                     return result;
                 }
 
-                result = expression.Compile();
+                var rewrittenExpression = Rewrite(expression, (MemberExpression)expression.Body);
+                result = ((Expression<Func<TModel, object>>)rewrittenExpression).Compile();
+
                 result = _simpleMemberAccessCache.GetOrAdd(memberExpression.Member, result);
+
                 return result;
             }
 
-            private static Func<TModel, TResult> CompileCapturedConstant(MemberExpression memberExpression, ConstantExpression constantExpression)
+            private static Func<TModel, object> CompileForChainedMemberAccess(Expression<Func<TModel, TResult>> expression)
+            {
+                // Input: () => m.Member1.Member
+                // Output: () => (m == null || m.Member1 == null) ? null : m.Member1.Member2
+                var rewrittenExpression = Rewrite(expression, (MemberExpression)expression.Body);
+                return ((Expression<Func<TModel, object>>)rewrittenExpression).Compile();
+            }
+
+            private static Func<TModel, object> CompileCapturedConstant(MemberExpression memberExpression, ConstantExpression constantExpression)
```

Comment:
I verified that this code path hasn't changed from before the cleanup. See https://github.com/aspnet/Mvc/blob/8b1bd343ba56152d2a7e677cf041dc9d62b4033f/src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/CachedExpressionCompiler.cs#L119-L139. Nothing to change here